File: src/gameplay/run_manager.py
import os
import random
from src.engine import event_bus
from src.gameplay.dictionary import Dictionary
from src.gameplay.scoring import check_word, calculate_word_score

class RunManager:
    def __init__(self, data_dir="data"):
        self.dictionary = Dictionary(data_dir)
        
        # Persistent Run State
        self.royalties = 4          # Starting money ($4)
        self.chapter = 1            # Starting Ante (Chapter 1)
        self.blind_index = 0        # 0: Small Blind, 1: Big Blind, 2: Boss Blind
        self.blind_type = "small"   # "small", "big", "boss"
        self.target_score = 500     # Score needed to win current blind
        
        self.tropes = []            # Active Tropes (max 5)
        self.edits = []             # Active consumable Edits (max 2)
        
        # Keyboard modifiers
        # For each letter 'a'-'z', tracks active mods: {"highlighter": bool, "coffee_ring": bool, "stapler": bool}
        self.keyboard_mods = {chr(c): {"highlighter": False, "coffee_ring": False, "stapler": False} for c in range(97, 123)}
        self.stapled_pairs = []     # List of sets/tuples of stapled keys, e.g., [('t', 'h')]
        
        # Style Guide Upgrade levels (default level 1)
        self.style_guides = {
            "Masterpiece": 1,
            "Jumble": 1,
            "Shot in the Dark": 1,
            "Total Rewrite": 1,
            "Standard Submission": 1
        }
        
        # Key status tracking for the current round (highest clue state discovered)
        # letter -> "green", "yellow", "grey", "empty"
        self.key_discoveries = {chr(c): "empty" for c in range(97, 123)}
        
        # Current Round State
        self.round_score = 0
        self.submissions_max = 4
        self.submissions_left = 4
        self.drafts_max = 2
        self.drafts_left = 2
        self.target_word = ""
        self.round_history = []     # List of dicts: {"word": str, "clues": list, "score": int, "is_draft": bool}
        
        # Boss Blind state
        self.boss_blind = None      # Current Boss Blind name: "Minimalist", "Plagiarist", "Ghostwriter", or None
        self.boss_pool = ["Minimalist", "Plagiarist", "Ghostwriter"]
        self.selected_boss = "Minimalist"  # The boss chosen for the current Chapter's end
        
        # Red Pen modifications
        self.red_pen_green_bonus = 0  # Permanent green letter chip bonus added by Red Pen
        
        # Choose initial boss blind for Chapter 1
        self.roll_boss_blind()

    def roll_boss_blind(self):
        """Rolls a random boss blind for the current chapter."""
        self.selected_boss = random.choice(self.boss_pool)

    def get_blind_name(self):
        if self.blind_index == 0:
            return "Small Blind"
        elif self.blind_index == 1:
            return "Big Blind"
        else:
            return f"Boss: {self.selected_boss}"

    def update_target_score(self):
        
        chapter_bases = [0, 500, 1500, 4000, 10000, 25000, 60000, 120000, 250000]
        base = chapter_bases[min(self.chapter, 8)]
        
        if self.blind_index == 0:
            self.target_score = base
            self.blind_type = "small"
            self.boss_blind = None
        elif self.blind_index == 1:
            self.target_score = int(base * 1.6)
            self.blind_type = "big"
            self.boss_blind = None
        else:
            self.target_score = int(base * 2.4)
            self.blind_type = "boss"
            self.boss_blind = self.selected_boss

        # Debuff from Ghostwriter Trope: Target Score increases by 1.5x
        if any(t.name == "The Ghostwriter" and t.is_debuff_active for t in self.tropes):
            self.target_score = int(self.target_score * 1.5)

    def start_round(self):
        """Initializes a new round, resetting submissions, drafts, and picking a target word."""
        self.round_score = 0
        self.round_history = []
        self.key_discoveries = {chr(c): "empty" for c in range(97, 123)}
        
        # Enforce Red Pen limits: lose 1 submission
        self.submissions_max = 4
        if any(t.name == "The Red Pen" and t.is_debuff_active for t in self.tropes):
            self.submissions_max = 3
            
        self.submissions_left = self.submissions_max
        self.drafts_left = self.drafts_max
        
        # Select target word
        # If boss is Minimalist, target word must be 4 letters long!
        if self.boss_blind == "Minimalist":
            # Generate a 4-letter word from target_words (truncate or select 4-letter)
            # Since standard wordlist contains 5-letter words, let's filter for 4-letter words or select a 4-letter target
            self.target_word = self.get_4_letter_target()
        else:
            self.target_word = self.dictionary.get_random_target()
            
        event_bus.bus.publish('ON_ROUND_START')

    def get_4_letter_target(self):
        """Generates or selects a 4-letter target word for the Minimalist Boss Blind."""
        # Let's extract 4-letter words from standard word list or create a pool
        words_4 = ["book", "page", "edit", "plot", "word", "desk", "read", "type", "inked", "bind"]
        # Strip to 4 letters to be safe
        words_4 = [w[:4].lower() for w in words_4]
        return random.choice(words_4)

    def submit_word(self, guess, is_draft=False):
        """
        Processes a guess, updates history, scores it (if not draft), and consumes resources.
        Returns a dict with clue results and score details.
        """
        guess = guess.lower().strip()
        
        # Validity checks
        if self.boss_blind == "Minimalist":
            if len(guess) != 4:
                return {"error": "Must submit a 4-letter word!"}
        else:
            if len(guess) != 5:
                return {"error": "Must submit a 5-letter word!"}
                
        # Plagiarism check (Hard Ban)
        # Checking if word was already submitted in this round as a real submission
        previous_submissions = {entry["word"] for entry in self.round_history if not entry.get("is_draft", False)}
        if not is_draft and guess in previous_submissions:
            # Plagiarized! Consumes a submission, awards 0 points, records in history.
            self.submissions_left -= 1
            clues = check_word(guess, self.target_word)
            result = {
                "score": 0,
                "chips": 0,
                "mult": 1.0,
                "x_mults": [],
                "clues": clues,
                "pattern": "Plagiarized"
            }
            self.round_history.append({
                "word": guess,
                "clues": clues,
                "is_draft": False,
                "score": 0,
                "is_plagiarized": True
            })
            return result
                
        # Wildcard check (Ghostwriter Trope)
        is_ghostwriter_active = any(t.name == "The Ghostwriter" for t in self.tropes)
        if is_ghostwriter_active:
            wildcard_count = guess.count('*')
            if wildcard_count > 2:
                return {"error": "Ghostwriter exploit! Max 2 wildcards allowed per submission."}
        
        # Validate against dictionary (if no wildcards or if we check letters)
        if not is_ghostwriter_active or "*" not in guess:
            # For Minimalist, we accept common 4 letter words or allow any guess for flexibility
            if self.boss_blind == "Minimalist":
                pass # Accept 4-letter guess
            elif not self.dictionary.is_valid_guess(guess):
                return {"error": "Not in word list!"}

        # Check stapled keys debuff
        # If one stapled key is played, the other MUST be played in the same word
        for pair in self.stapled_pairs:
            letter1, letter2 = pair
            has1 = letter1 in guess
            has2 = letter2 in guess
            if (has1 and not has2) or (has2 and not has1):
                return {"error": f"Stapled keys! Must play '{letter1.upper()}' and '{letter2.upper()}' together."}

        # Check Plot Twist trope debuff
        if any(t.name == "The Plot Twist" and t.is_debuff_active for t in self.tropes):
            if 'e' in guess:
                return {"error": "Plot Twist debuff! Cannot play words containing 'E'."}

        # Check Purple Prose trope debuff
        if any(t.name == "Purple Prose" and t.is_debuff_active for t in self.tropes):
            vowels = sum(1 for char in guess if char in "aeiou")
            if vowels < 2:
                return {"error": "Purple Prose debuff! Must use at least 2 vowels."}

        # Check Red Pen trope debuff
        # Grey letters are permanently removed from the keyboard
        if any(t.name == "The Red Pen" and t.is_debuff_active for t in self.tropes):
            # Find which letters were previously marked grey in this round/run
            # We can check if the guess contains any letter that was marked grey
            # Let's track removed letters in self.keyboard_mods or run state.
            for char in guess:
                if self.keyboard_mods.get(char, {}).get("removed", False):
                    return {"error": f"Red Pen debuff! Letter '{char.upper()}' has been removed."}

        # Handle Boss Blind: The Plagiarist
        # Target word is guaranteed to be an anagram of the very first word submitted.
        if self.boss_blind == "Plagiarist" and len(self.round_history) == 0 and not is_draft:
            # We set the target word to an anagram of the first guess!
            self.target_word = self.get_plagiarist_anagram(guess)

        # Run scoring/comparison
        # For wildcard matches, replace '*' in guess with matching letter in target
        eval_guess = list(guess)
        if is_ghostwriter_active:
            for idx in range(len(guess)):
                if idx < len(self.target_word) and guess[idx] == '*':
                    eval_guess[idx] = self.target_word[idx]
        eval_guess_str = "".join(eval_guess)

        clues = check_word(eval_guess_str, self.target_word)
        
        # Apply Red Pen: grey letters turn removed, and trigger permanent green value boost
        is_red_pen_active = any(t.name == "The Red Pen" for t in self.tropes)
        if is_red_pen_active and not is_draft:
            for idx, clue in enumerate(clues):
                char = guess[idx]
                if clue == 'grey' and char.isalpha():
                    if not self.keyboard_mods[char].get("removed", False):
                        self.keyboard_mods[char]["removed"] = True
                        self.red_pen_green_bonus += 10

        # Calculate scores
        if is_draft:
            self.drafts_left -= 1
            result = {
                "score": 0,
                "chips": 0,
                "mult": 1.0,
                "x_mults": [],
                "clues": clues,
                "pattern": "Drafted"
            }
        else:
            self.submissions_left -= 1
            
            # Temporary hook for Red Pen green bonus
            # We can register Red Pen listener to ON_LETTER_SCORED, or apply it here:
            result = calculate_word_score(eval_guess_str, self.target_word, self.style_guides, self.keyboard_mods, boss_blind=self.boss_blind)
            
            # Apply Red Pen green letter boost if green
            if self.red_pen_green_bonus > 0:
                for idx, clue in enumerate(clues):
                    if clue == 'green':
                        result["chips"] += self.red_pen_green_bonus
                        # Re-calculate score total
                        total_mult = result["mult"]
                        for xm in result["x_mults"]:
                            total_mult *= xm
                        result["score"] = int(result["chips"] * total_mult)
            
            self.round_score += result["score"]
            
        # Redact colors if Coffee Ring is active on a letter
        final_clues = list(clues)
        for idx, char in enumerate(guess):
            if self.keyboard_mods.get(char, {}).get("coffee_ring", False):
                final_clues[idx] = "redacted" # clue is hidden

        history_entry = {
            "word": guess,
            "clues": final_clues,
            "score": result["score"],
            "is_draft": is_draft
        }
        self.round_history.append(history_entry)
        
        # Publish submission events
        if is_draft:
            event_bus.bus.publish('ON_WORD_DRAFTED', guess, final_clues)
        else:
            event_bus.bus.publish('ON_WORD_SUBMITTED', guess, final_clues, result["score"])
            
        return result

    def get_plagiarist_anagram(self, first_guess):
        """Returns a valid anagram of the first guess to act as the new target word."""
        # Find all words in dictionary targets that are anagrams of first_guess
        sorted_guess = sorted(first_guess)
        candidates = [w for w in self.dictionary.target_words if sorted(w) == sorted_guess and w != first_guess]
        if not candidates:
            # Check guess list
            candidates = [w for w in self.dictionary.allowed_guesses if sorted(w) == sorted_guess and w != first_guess]
        
        if candidates:
            return random.choice(candidates)
        else:
            # Fallback: shuffle the first guess to create an anagram
            chars = list(first_guess)
            random.shuffle(chars)
            return "".join(chars)

    def check_round_end(self):
        """Checks if the round is over. Returns 'win', 'lose', or 'continue'."""
        if self.round_score >= self.target_score:
            # Won round! Gain royalties: base $3 + $1 per remaining submission
            reward = 3 + self.submissions_left
            self.royalties += reward
            event_bus.bus.publish('ON_STAGE_WON', reward)
            return "win"
        elif self.submissions_left <= 0:
            # Lost round! Fired!
            event_bus.bus.publish('ON_STAGE_LOST')
            return "lose"
        return "continue"

    def advance_blind(self):
        """Advances to the next blind. If chapter boss is beaten, moves to next chapter."""
        self.blind_index += 1
        if self.blind_index > 2:
            self.chapter += 1
            self.blind_index = 0
            # Reset any temporary debuffs like Red Pen removed keys for the new chapter
            for char in self.keyboard_mods:
                self.keyboard_mods[char]["removed"] = False
            self.roll_boss_blind()
            
        self.update_target_score()

File: love2d/src/gameplay/run_manager.lua
local event_bus = require("src/engine/event_bus")
local Dictionary = require("src/gameplay/dictionary")
local scoring = require("src/gameplay/scoring")

local RunManager = {}
RunManager.__index = RunManager

function RunManager.new(data_dir)
    local self = setmetatable({}, RunManager)
    self.dictionary = Dictionary.new(data_dir)
    
    -- Persistent Run State
    self.royalties = 4
    self.chapter = 1
    self.assignment_index = 0
    self.assignment_type = "small"
    self.target_score = 500
    
    self.tropes = {}
    self.edits = {}
    
    -- Keyboard modifiers
    self.keyboard_mods = {}
    for c = 97, 122 do
        local char = string.char(c)
        self.keyboard_mods[char] = {
            highlighter = false,
            coffee_ring = false,
            stapler = false,
            removed = false,
            correction_tape = false
        }
    end
    self.stapled_pairs = {}
    
    local style_guides_data = require("src/content/style_guides").STYLE_GUIDES_DATA
    self.style_guides = {
        ["Standard Submission"] = 1
    }
    for pat, _ in pairs(style_guides_data) do
        self.style_guides[pat] = 1
    end
    
    self.key_discoveries = {}
    for c = 97, 122 do
        self.key_discoveries[string.char(c)] = "empty"
    end
    
    -- Current Round State
    self.round_score = 0
    self.submissions_max = 4
    self.submissions_left = 4
    self.drafts_max = 2
    self.drafts_left = 2
    self.target_word = ""
    self.round_history = {}
    
    -- Boss Assignment state
    self.boss_assignment = nil
    self.boss_pool = {"Minimalist", "Plagiarist", "Ghostwriter"}
    self.selected_boss = "Minimalist"
    
    self.red_pen_green_bonus = 0
    
    self:roll_boss_assignment()
    self:update_target_score()
    
    return self
end

function RunManager:roll_boss_assignment()
    local idx = love.math.random(1, #self.boss_pool)
    self.selected_boss = self.boss_pool[idx]
end

function RunManager:get_assignment_name()
    if self.assignment_index == 0 then
        return "Small Assignment"
    elseif self.assignment_index == 1 then
        return "Big Assignment"
    else
        return "Boss: " .. self.selected_boss
    end
end

function RunManager:update_target_score()
    local chapter_bases = {0, 500, 1500, 4000, 10000, 25000, 60000, 120000, 250000}
    local base_idx = math.min(self.chapter, 8) + 1 -- chapter 1 maps to chapter_bases[2]
    local base = chapter_bases[base_idx] or 250000
    
    if self.assignment_index == 0 then
        self.target_score = base
        self.assignment_type = "small"
        self.boss_assignment = nil
    elseif self.assignment_index == 1 then
        self.target_score = math.floor(base * 1.6)
        self.assignment_type = "big"
        self.boss_assignment = nil
    else
        self.target_score = math.floor(base * 2.4)
        self.assignment_type = "boss"
        self.boss_assignment = self.selected_boss
    end
    
    -- Ghostwriter trope target modifier
    local has_ghostwriter = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Ghostwriter" and t.is_debuff_active then
            has_ghostwriter = true
            break
        end
    end
    if has_ghostwriter then
        self.target_score = math.floor(self.target_score * 1.5)
    end
end

function RunManager:start_round()
    self.round_score = 0
    self.round_history = {}
    for c = 97, 122 do
        self.key_discoveries[string.char(c)] = "empty"
    end
    
    self.submissions_max = 4
    local has_red_pen = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Red Pen" and t.is_debuff_active then
            has_red_pen = true
            break
        end
    end
    if has_red_pen then
        self.submissions_max = 3
    end
    
    self.submissions_left = self.submissions_max
    self.drafts_left = self.drafts_max
    
    if self.boss_assignment == "Minimalist" then
        self.target_word = self:get_4_letter_target()
    else
        self.target_word = self.dictionary:get_random_target()
    end
    
    event_bus.bus:publish('ON_ROUND_START')
end

function RunManager:get_4_letter_target()
    local words_4 = {"book", "page", "edit", "plot", "word", "desk", "read", "type", "inks", "bind"}
    local idx = love.math.random(1, #words_4)
    return words_4[idx]:sub(1, 4):lower()
end

local function sort_string(s)
    local chars = {}
    for i = 1, #s do
        table.insert(chars, s:sub(i, i))
    end
    table.sort(chars)
    return table.concat(chars)
end

function RunManager:get_plagiarist_anagram(first_guess)
    local sorted_guess = sort_string(first_guess)
    local candidates = {}
    
    for _, w in ipairs(self.dictionary.target_words) do
        if sort_string(w) == sorted_guess and w ~= first_guess then
            table.insert(candidates, w)
        end
    end
    
    if #candidates == 0 then
        for w, _ in pairs(self.dictionary.allowed_guesses) do
            if sort_string(w) == sorted_guess and w ~= first_guess then
                table.insert(candidates, w)
            end
        end
    end
    
    if #candidates > 0 then
        return candidates[love.math.random(1, #candidates)]
    else
        local chars = {}
        for i = 1, #first_guess do
            table.insert(chars, first_guess:sub(i, i))
        end
        -- Shuffle chars
        for i = #chars, 2, -1 do
            local j = love.math.random(1, i)
            chars[i], chars[j] = chars[j], chars[i]
        end
        return table.concat(chars)
    end
end

function RunManager:submit_word(guess, is_draft)
    guess = guess:lower():match("^%s*(.-)%s*$")
    is_draft = is_draft or false
    
    if self.boss_assignment == "Minimalist" then
        if #guess ~= 4 then
            return {error = "Must submit a 4-letter word!"}
        end
    else
        if #guess ~= 5 then
            return {error = "Must submit a 5-letter word!"}
        end
    end
    
    -- Plagiarism check
    local previous_submissions = {}
    for _, entry in ipairs(self.round_history) do
        if not entry.is_draft then
            previous_submissions[entry.word] = true
        end
    end
    if not is_draft and previous_submissions[guess] then
        self.submissions_left = self.submissions_left - 1
        local clues = scoring.check_word(guess, self.target_word)
        local result = {
            score = 0,
            words = 0,
            hype = 1.0,
            x_hypes = {},
            clues = clues,
            pattern = "Plagiarized"
        }
        table.insert(self.round_history, {
            word = guess,
            clues = clues,
            is_draft = false,
            score = 0,
            is_plagiarized = true
        })
        return result
    end
    
    local is_ghostwriter_active = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Ghostwriter" then
            is_ghostwriter_active = true
            break
        end
    end
    
    if is_ghostwriter_active then
        local _, wildcards = guess:gsub("%*", "")
        if wildcards > 2 then
            return {error = "Ghostwriter exploit! Max 2 wildcards allowed per submission."}
        end
    end
    
    if not is_ghostwriter_active or not guess:find("%*") then
        if self.boss_assignment == "Minimalist" then
            -- Accept all 4 letter words
        elseif not self.dictionary:is_valid_guess(guess) then
            return {error = "Not in word list!"}
        end
    end
    
    for _, pair in ipairs(self.stapled_pairs) do
        local letter1, letter2 = pair[1], pair[2]
        local has1 = guess:find(letter1, 1, true) ~= nil
        local has2 = guess:find(letter2, 1, true) ~= nil
        if (has1 and not has2) or (has2 and not has1) then
            return {error = "Stapled keys! Must play '" .. letter1:upper() .. "' and '" .. letter2:upper() .. "' together."}
        end
    end
    
    local has_plot_twist_debuff = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Plot Twist" and t.is_debuff_active then
            has_plot_twist_debuff = true
            break
        end
    end
    if has_plot_twist_debuff and guess:find("e", 1, true) then
        return {error = "Plot Twist debuff! Cannot play words containing 'E'."}
    end
    
    local has_purple_prose_debuff = false
    for _, t in ipairs(self.tropes) do
        if t.name == "Purple Prose" and t.is_debuff_active then
            has_purple_prose_debuff = true
            break
        end
    end
    if has_purple_prose_debuff then
        local vowels = 0
        for i = 1, #guess do
            local char = guess:sub(i, i)
            if char:find("[aeiou]") then
                vowels = vowels + 1
            end
        end
        if vowels < 2 then
            return {error = "Purple Prose debuff! Must use at least 2 vowels."}
        end
    end
    
    local has_red_pen_debuff = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Red Pen" and t.is_debuff_active then
            has_red_pen_debuff = true
            break
        end
    end
    if has_red_pen_debuff then
        for i = 1, #guess do
            local char = guess:sub(i, i)
            if self.keyboard_mods[char] and self.keyboard_mods[char].removed then
                return {error = "Red Pen debuff! Letter '" .. char:upper() .. "' has been removed."}
            end
        end
    end
    
    if self.boss_assignment == "Plagiarist" and #self.round_history == 0 and not is_draft then
        self.target_word = self:get_plagiarist_anagram(guess)
    end
    
    local eval_guess_chars = {}
    for i = 1, #guess do
        local c = guess:sub(i, i)
        if is_ghostwriter_active and c == "*" and i <= #self.target_word then
            table.insert(eval_guess_chars, self.target_word:sub(i, i))
        else
            table.insert(eval_guess_chars, c)
        end
    end
    local eval_guess_str = table.concat(eval_guess_chars)
    
    local clues = scoring.check_word(eval_guess_str, self.target_word)
    
    local has_red_pen = false
    for _, t in ipairs(self.tropes) do
        if t.name == "The Red Pen" then
            has_red_pen = true
            break
        end
    end
    if has_red_pen and not is_draft then
        for idx = 1, #clues do
            local clue = clues[idx]
            local char = guess:sub(idx, idx)
            if clue == 'grey' and char:match("%a") then
                if self.keyboard_mods[char] and not self.keyboard_mods[char].removed then
                    self.keyboard_mods[char].removed = true
                    self.red_pen_green_bonus = self.red_pen_green_bonus + 10
                end
            end
        end
    end
    
    local result
    if is_draft then
        self.drafts_left = self.drafts_left - 1
        result = {
            score = 0,
            words = 0,
            hype = 1.0,
            x_hypes = {},
            clues = clues,
            pattern = "Drafted"
        }
    else
        self.submissions_left = self.submissions_left - 1
        
        result = scoring.calculate_word_score(eval_guess_str, self.target_word, self.style_guides, self.keyboard_mods, self.boss_assignment)
        
        if self.red_pen_green_bonus > 0 then
            for idx = 1, #clues do
                if clues[idx] == 'green' then
                    result.words = result.words + self.red_pen_green_bonus
                    local total_hype = result.hype
                    for _, xm in ipairs(result.x_hypes) do
                        total_hype = total_hype * xm
                    end
                    result.score = math.floor(result.words * total_hype)
                end
            end
        end
        
        self.round_score = self.round_score + result.score
    end
    
    local final_clues = {}
    for idx = 1, #guess do
        local char = guess:sub(idx, idx)
        if self.keyboard_mods[char] and self.keyboard_mods[char].coffee_ring then
            final_clues[idx] = "redacted"
        else
            final_clues[idx] = clues[idx]
        end
    end
    
    local history_entry = {
        word = guess,
        clues = final_clues,
        score = result.score,
        is_draft = is_draft
    }
    table.insert(self.round_history, history_entry)
    
    if is_draft then
        event_bus.bus:publish('ON_WORD_DRAFTED', guess, final_clues)
    else
        event_bus.bus:publish('ON_WORD_SUBMITTED', guess, final_clues, result.score)
    end
    
    return result
end

function RunManager:check_round_end()
    if self.round_score >= self.target_score then
        local reward = 3 + self.submissions_left
        self.royalties = self.royalties + reward
        event_bus.bus:publish('ON_STAGE_WON', reward)
        return "win"
    elseif self.submissions_left <= 0 then
        event_bus.bus:publish('ON_STAGE_LOST')
        return "lose"
    end
    return "continue"
end

function RunManager:advance_assignment()
    self.assignment_index = self.assignment_index + 1
    if self.assignment_index > 2 then
        self.chapter = self.chapter + 1
        self.assignment_index = 0
        for _, mod in pairs(self.keyboard_mods) do
            mod.removed = false
        end
        self:roll_boss_assignment()
    end
    self:update_target_score()
end

return RunManager
